ragna/source_storages/_qdrant.py

```diff
@@ -0,0 +1,249 @@
+                    ):
+                        continue
+
+                    corpus_metadata[key].add(value)
+
+            metadata[corpus_name] = {
+                key: ({type(value).__name__ for value in values}.pop(), sorted(values))
+                for key, values in corpus_metadata.items()
+            }
+
+        return metadata
+
+    def store(
+        self,
+        corpus_name: str,
+        documents: list[Document],
+        *,
+        chunk_size: int = 500,
+        chunk_overlap: int = 250,
+    ) -> None:
+        from qdrant_client import models
+
+        self._ensure_table(corpus_name, create=True)
+
+        points = []
+        for document in documents:
+            for chunk in self._chunk_pages(
+                document.extract_pages(),
+                chunk_size=chunk_size,
+                chunk_overlap=chunk_overlap,
+            ):
+                points.append(
+                    models.PointStruct(
+                        id=str(uuid.uuid4()),
+                        vector=self._embedding_function([chunk.text])[0],
+                        payload={
+                            "document_id": str(document.id),
+                            "document_name": document.name,
+                            **document.metadata,
+                            "__page_numbers__": self._page_numbers_to_str(
+                                chunk.page_numbers
+                            ),
+                            "__num_tokens__": chunk.num_tokens,
+                            self.DOC_CONTENT_KEY: chunk.text,
+                        },
+                    )
+                )
+
+        self._client.upsert(collection_name=corpus_name, points=points)
+
+    def _build_condition(self, operator, key, value):
+        from qdrant_client import models
+
+        if operator == MetadataOperator.EQ:
+            return models.FieldCondition(key=key, match=models.MatchValue(value=value))
+        elif operator == MetadataOperator.LT:
+            return models.FieldCondition(key=key, range=models.Range(lt=value))
+        elif operator == MetadataOperator.LE:
+            return models.FieldCondition(key=key, range=models.Range(lte=value))
+        elif operator == MetadataOperator.GT:
+            return models.FieldCondition(key=key, range=models.Range(gt=value))
+        elif operator == MetadataOperator.GE:
+            return models.FieldCondition(key=key, range=models.Range(gte=value))
+        elif operator == MetadataOperator.IN:
+            return models.FieldCondition(key=key, match=models.MatchAny(any=value))
+        elif operator in {MetadataOperator.NE, MetadataOperator.NOT_IN}:
+            except_value = [value] if operator == MetadataOperator.NE else value
+            return models.FieldCondition(
+                key=key, match=models.MatchExcept(**{"except": except_value})
+            )
+        else:
+            raise ValueError(f"Unsupported operator: {operator}")
```

Comment:
This only applies to the `RAW` operator, correct? Looking at the documentation, there are plenty of other operators. Is it possible to take a JSON filter even when running in Python? If so this is probably very similar to what Chroma is doing:

https://github.com/Quansight/ragna/blob/71acf83d4114514c56e298f7acc431551933e522/ragna/source_storages/_chroma.py#L153-L154